torch/_inductor/scheduler.py

```diff
@@ -2416,21 +2401,38 @@ def can_fuse_vertical(
                 why("intermediate nodes between node1 & node2")
                 return False
 
-        # similar to can_inplace, if we are going to fuse a write subsequent to a read
-        # require that the indexing and size is the same
-        for write in node2.read_writes.writes:
-            if not isinstance(write, MemoryDep):
-                continue
-            for read in node1.read_writes.reads:
-                if write.name != self.mutation_renames.get(read.name, read.name):
-                    continue
+        return True
 
-                # bail on StarDep
-                if not self.fusable_read_and_write(read, write):
-                    why("fusing a write into a read with different indexing formula")
-                    return False
+    def fusable_weak_dep(
+        self, weak_dep: WeakDep, node1: BaseSchedulerNode, node2: BaseSchedulerNode
+    ) -> bool:
+        # A weak dep can be fused if and only if the fused operation acts inplace
+        # on the buffer being mutated. i.e. the same index is being read then mutated
+        mutating_writes = [
+            write
+            for write in node2.read_writes.writes
+            if write.name == weak_dep.mutating_buf
+        ]
+        if len(mutating_writes) != 1:
+            return False
```

Comment:
What's an example when len > 1?

Reply:
`upsample_backward` with multiple different accumulation mutations on the input's gradient.

Reply:
You mean like in https://github.com/pytorch/pytorch/pull/123222#issuecomment-2040045663?
But in that case it's possible to fuse them, no? I don't think I understand.

Reply:
I suppose we could check that all the mutation writes happen at a single index and allow it in that case. However it would be a bit unusual to have multiple atomic_add calls to exactly the same index.

The case we can't fuse is where we mutate multiple indices in the same buffer, which I expect to be the common case

Reply:
I still don't understand.
In the codegen linked above, we have
```python
    tl.atomic_add(out_ptr0 + (tmp12 + (500*tmp6) + (2500000*x2)), tmp21, xmask)
    tl.atomic_add(out_ptr0 + (tmp25 + (500*tmp6) + (2500000*x2)), tmp26, xmask)
    tl.atomic_add(out_ptr0 + (tmp12 + (500*tmp29) + (2500000*x2)), tmp21, xmask)
    tl.atomic_add(out_ptr0 + (tmp25 + (500*tmp29) + (2500000*x2)), tmp26, xmask)
```
All these are atomic_adds that write to different addresses, but it's completely fine to fuse them.

I guess it would be problematic if:
- They were writing to different addresses
- They were in a for loop that loaded the tensor in chunks (similar to non-persistent reductions) and the loads and stores didn't play well with each other
Is this the case you are thinking about?

Reply:
Yes multiple writes to different indexes are find on their own, but we can't fuse them with reads. Say for example you have:
```
tmp0 = tl.load(in_out_ptr0 + x0, xmask)
tl.atomic_add(in_out_ptr0 + x0, tmp0, xmask)
tl.atomic_add(in_out_ptr0 + x0 + 1, tmp0, (x0 + 1 < xnumel) & xmask)
```

We have a race condition where `tmp0` could be read either before or after the neighbouring thread has modified the underlying value.

Reply:
Ah, that's completely fair.
In a future PR, we could still fuse these by adding some explicit ordering by turning the load into an `atomic_add(..., 0)` and doing an acquire / release.